### PR TITLE
Laboratory Work #1, Sophya Perepelkina - 20FPL3

### PR DESCRIPTION
**Pull request by SophyaPerepelkina**

### CODE REVIEW

File: scrapper.py [hunk 1]
```diff
@@ -1,7 +1,12 @@
 """
 Scrapper implementation
 """
-
+import json
+import os
+import requests
+from bs4 import BeautifulSoup
+from constants import ASSETS_PATH, CRAWLER_CONFIG_PATH
+#from core_utils.article import Article
```
Comment (marina-kaz) [inline scrapper.py:9]: why is this commented?
we do need Article :)

File: scrapper.py [hunk 2]
```diff
@@ -26,36 +31,66 @@ class Crawler:
...
 
     def find_articles(self):
         """
         Finds articles
         """
-        pass
+        for urls in self.seed_urls:
+            response = requests.get(urls)
+            response.encoding = 'utf-16'
+            with open('index.html', 'w', encoding='utf-16') as file:
```
Comment (marina-kaz) [inline scrapper.py:52]: This code is going to create a html file in the directory where scrapper.py is located.
We do not need to save all the visited pages as html.
Refer to `docs/datset.md` to see where and how data must be stored.

**Review by marina-kaz — CHANGES_REQUESTED** [review]
There is not much time left so I advise to speed up a little :)

**Review by marina-kaz — CHANGES_REQUESTED** [review]
There aren't any significant changes. Time to speed up!

File: scrapper.py [hunk 3]
```diff
@@ -26,38 +34,120 @@ class Crawler:
...
     """
     def __init__(self, seed_urls, max_articles: int):
-        pass
+        self.seed_urls = seed_urls
+        self.max_articles = max_articles
+        self.urls = []
 
     def _extract_url(self, article_bs):
-        pass
+        self.urls.append('https://magadanpravda.ru/' + article_bs.find('a')['href'])
```
Comment (marina-kaz) [inline scrapper.py:42]: do you only add one URL from each seed URL? 
also, before adding a new link, it is better to check that this link has not yet been added as sources often place the same links over and over

File: test.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import requests
```
[the remark this answers is not in the document]
Comment (marina-kaz) [reply test.py]: remove this file if you do not need it

File: scrapper.py [hunk 5]
```diff
@@ -26,38 +34,120 @@ class Crawler:
...
+    if not isinstance(max_articles, int) or max_articles <= 0:
+        raise IncorrectNumberOfArticlesError
+    if not isinstance(seed_urls, list) or not seed_urls:
+        raise IncorrectURLError
+    if max_articles > 200:
+        raise NumberOfArticlesOutOfRangeError
+    for urls in seed_urls:
+        if urls[0:8] != 'https://' and urls[0:7] != 'http://':
+            raise IncorrectURLError
+    return seed_urls, max_articles
```
Comment (marina-kaz) [inline scrapper.py:140]: what if seed_urls is an empty list? would that still be considered correct, according to this function?

**Review by marina-kaz — CHANGES_REQUESTED** [review]
A little polishing would help

File: scrapper.py [hunk 6]
```diff
@@ -26,36 +31,66 @@ class Crawler:
...
 def validate_config(crawler_path):
     """
     Validates given config
     """
-    pass
+    with open(crawler_path, 'r', encoding='utf-8') as file:
+        config = json.load(file)
+    seed_urls = config['seed_urls']
+    max_articles = config['total_articles_to_find_and_parse']
+    if not isinstance(max_articles, int) or max_articles <= 0:
```
[the remark this answers is not in the document]
Comment (SophyaPerepelkina) [reply scrapper.py:131]: Fixed this issue

File: scrapper.py (hunk 5 above)
Comment (SophyaPerepelkina) [reply scrapper.py:140]: Umm... isn't it checked by "if not seed_urls"?

Comment (marina-kaz) [reply scrapper.py:144]: True :)

File: scrapper.py (hunk 3 above)
Comment (SophyaPerepelkina) [reply scrapper.py:42]: I checked, I do, indeed, collect unique URLs

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to finish Crawler and embark on Pipeline
Also, please, delete extra files: `test.py`

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Nothing to review just yet

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to implement pipeline

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Not much to review

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Nothing to review

**Pull request closed without merging**